beefy-gadget/src/worker.rs

```diff
@@ -55,12 +63,101 @@ where
 	<<B::Header as Header>::Hashing as Hash>::hash(b"beefy")
 }
 
+/// Allows messages from last `MAX_LIVE_GOSSIP_ROUNDS` to flow, everything else gets
+/// rejected/expired. All messaging is handled in a single global topic.
+pub struct BeefyGossipValidator<B, P>
+where
+	B: Block,
+{
+	topic: B::Hash,
+	live_rounds: RwLock<Vec<NumberFor<B>>>,
+	_pair: PhantomData<P>,
+}
+
+impl<B, P> BeefyGossipValidator<B, P>
+where
+	B: Block,
+{
+	pub fn new() -> BeefyGossipValidator<B, P> {
+		BeefyGossipValidator {
+			topic: topic::<B>(),
+			live_rounds: RwLock::new(Vec::new()),
+			_pair: PhantomData,
+		}
+	}
+
+	fn note_round(&self, round: NumberFor<B>) {
+		let mut live_rounds = self.live_rounds.write();
+
+		// NOTE: ideally we'd use a VecDeque here, but currently binary search is only available on
+		// nightly for `VecDeque`.
+		while live_rounds.len() > MAX_LIVE_GOSSIP_ROUNDS {
+			let _ = live_rounds.remove(0);
+		}
+
+		if let Some(idx) = live_rounds.binary_search(&round).err() {
+			live_rounds.insert(idx, round);
+		}
+	}
+
+	fn is_live(live_rounds: &[NumberFor<B>], round: NumberFor<B>) -> bool {
+		live_rounds.binary_search(&round).is_ok()
+	}
+}
+
+impl<B, P> GossipValidator<B> for BeefyGossipValidator<B, P>
+where
+	B: Block,
+	P: Pair,
+	P::Public: Decode,
+	P::Signature: Decode,
+{
+	fn validate(
+		&self,
+		_context: &mut dyn GossipValidatorContext<B>,
+		_sender: &sc_network::PeerId,
+		mut data: &[u8],
+	) -> GossipValidationResult<B::Hash> {
+		if VoteMessage::<MmrRootHash, NumberFor<B>, P::Public, P::Signature>::decode(&mut data).is_ok() {
```

Comment:
I suppose this happens after `message_allowed` returns true, right? Shouldn't we actually verify signatures here already?

Reply:
This is the method that's called first by the gossip layer when we receive the message for the first time (and only called once while we still hold on to this message). So yes, this would be a good place to verify signatures as we can immediately discard the message if the signature is invalid. We are currently not checking signatures though AFAIK.